include/common/aglResCommon.h

```diff
@@ -0,0 +1,185 @@
+    void setEndianResolved() { ref().mEndian = 1 - ref().mEndian; }
+
+template <typename DataType>
+struct ResArrayData {
+    s32 mSize;
+    u32 mNum;
+    // DataType mData[];
+
+    using ElemType = DataType;
+};
+
+template <typename Type>
+class ResArray : public ResCommon<ResArrayData<typename Type::DataType>> {
+public:
+    using ElemType = Type;
+    using ElemDataType = typename Type::DataType;
+    using DataType = typename ResArray<Type>::DataType;
+    using Base = ResCommon<DataType>;
+
+    using ResCommon<DataType>::ResCommon;
+
+public:
+    class iterator {
+    public:
+        iterator(s32 index, ElemDataType* elem) : mIndex(index), mElem(elem) {}
+
+        friend bool operator==(const iterator& lhs, const iterator& rhs) {
+            return lhs.mIndex == rhs.mIndex;
+        }
+
+        friend bool operator!=(const iterator& lhs, const iterator& rhs) {
+            return lhs.mIndex != rhs.mIndex;
+        }
+
+        iterator& operator++() {
+            ++mIndex;
+            mElem = (ElemDataType*)((uintptr_t)mElem + Type(mElem).ref().mSize);
+            return *this;
+        }
+
+        ElemDataType& operator*() const { return *mElem; }
+        ElemDataType* operator->() const { return mElem; }
+        s32 getIndex() const { return mIndex; }
+
+    private:
+        s32 mIndex;
+        ElemDataType* mElem;
+    };
+
+    class constIterator {
+    public:
+        constIterator(s32 index, const ElemDataType* elem) : mIndex(index), mElem(elem) {}
+
+        friend bool operator==(const constIterator& lhs, const constIterator& rhs) {
+            return lhs.mIndex == rhs.mIndex;
+        }
+
+        friend bool operator!=(const constIterator& lhs, const constIterator& rhs) {
+            return lhs.mIndex != rhs.mIndex;
+        }
+
+        constIterator& operator++() {
+            ++mIndex;
+            mElem = (const ElemDataType*)((uintptr_t)mElem + Type(mElem).ref().mSize);
+            return *this;
+        }
+
+        const ElemDataType& operator*() const { return *mElem; }
+        const ElemDataType* operator->() const { return mElem; }
+        s32 getIndex() const { return mIndex; }
+
+    private:
+        s32 mIndex;
+        const ElemDataType* mElem;
+    };
+
+public:
+    iterator begin() { return iterator(0, (ElemDataType*)(Base::ptr() + 1)); }
+    constIterator begin() const { return constIterator(0, (const ElemDataType*)(Base::ptr() + 1)); }
+    constIterator constBegin() const {
+        return constIterator(0, (const ElemDataType*)(Base::ptr() + 1));
+    }
+
+    iterator end() { return iterator(getNum(), nullptr); }
+    constIterator end() const { return constIterator(getNum(), nullptr); }
+    constIterator constEnd() const { return constIterator(getNum(), nullptr); }
+
+public:
+    u32 getNum() const { return Base::ref().mNum; }
+
+    ElemType get(s32 n) const {
+        SEAD_ASSERT(0 <= n && n <= static_cast<int>(this->getNum()));
```

Comment:
The spaces here were part of the original assertion message and it wouldn't match anymore now.

Reply:
What are you trying to tell me here? Which spaces are you referring to, and why would code formatting affect code matching?

Reply:
https://github.com/open-ead/agl/pull/15/commits/c67cf4195341871dd0828688f64795c6cefa45ee

(Since we are not building debug builds, I don't think matching is important for them, but imo it would be better if we could stay accurate to the original assertion message.)

Reply:
Huh? Why would those spaces there matter for the code generated at all? That's just ... useless spacing, or how would that give a different code?

Reply:
As I said several times, it would generate a different assertion message in debug builds.

Reply:
Alright, added back in and added tags to ignore it for formatting to stop `clang-format` from complaining.